### PR TITLE
[Spark] Don't commit empty transactions on Writes #1933

### LINKED ISSUE #1933: [Feature Request] Don't write empty transactions to the log
## Feature request
As an optimization, don't write empty transactions to the log.

#### Which Delta project/connector is this regarding?
- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

### Overview

If the transaction is a no-op / empty, then no transaction should be committed.

### Motivation

I have a streaming job that ingests sensor data into a few dozen delta tables. Some of this data is beyond our ingestion window and goes to an "expired" table. For most of the batches this ends up with an empty dataframe going into the call to `df.write`. I noticed on our expired records table the majority of the log entries per day are completely empty.

### Further details

At least in our case I cannot see the utility in the empty transaction being recorded, I guess I could see in other cases where the transaction log entry would serve as a high watermark of sorts to indicate liveness. Perhaps a new flag could be added to omit empty transactions.

### Willingness to contribute

The Delta Lake Community encourages new feature contributions. Would you or another member of your organization be willing to contribute an implementation of this feature?

- [X] Yes. I can contribute this feature independently.
- [ ] Yes. I would be willing to contribute this feature with guidance from the Delta Lake community.
- [ ] No. I cannot contribute this feature at this time.

### PR DESCRIPTION
**Pull request by watfordkcf**
#### Which Delta project/connector is this regarding?
- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

This PR makes the write command behave similarly to delete, update, and merge, in that it will not write an empty commit to the transaction log.

Resolves #1933 

## How was this patch tested?

Working on testing this locally, `sbt test` hangs on the `hiveTez` suite currently.

## Does this PR introduce _any_ user-facing changes?

Yes, if the user consumes the delta transaction log and would expect an empty write transaction to include a record in the log.

### CODE REVIEW

**Comment by tdas** [comment]
@prakharjain09 could you take a look at this?

**Comment by felipepessoto** [comment]
There is a scenario where empty commit can be useful, when you need to add a "time marker" (Idk how to call it). For example, if you have sporadic writes, and want to read the table as it is Today 00:00:00. If there is no commit after that timestamp, you can't time travel, the workaround is creating an empty commit.

Do we have any other workaround to that scenario?

**Comment by watfordkcf** [comment]
If the default for `delta.skipRecordingEmptyCommits` is `true` then I do not believe so.

**Comment by felipepessoto** [comment]
How are the tests working if skipRecordingEmptyCommits  says `This only works when table is using SnapshotIsolation or Serializable Isolation Mode`. And default is WriteSerializable?

The skipRecordingEmptyCommits description is incorrect? Or is it a bug?

**Comment by watfordkcf** [comment]
The default is `Serializable` or is this test wrong?
https://github.com/delta-io/delta/blob/c78daeff2062ce3c9d8c159301add1848f25c809/spark/src/test/scala/org/apache/spark/sql/delta/DeltaConfigSuite.scala#L179-L190

**Comment by felipepessoto** [comment]
You are right, I'm looking the incorrect place:

https://github.com/delta-io/delta/blob/c78daeff2062ce3c9d8c159301add1848f25c809/spark/src/main/scala/org/apache/spark/sql/delta/isolationLevels.scala#L73C1-L75C34

WriteSerializable is not even implemented yet: https://github.com/delta-io/delta/pull/1262

**Comment by watfordkcf** [comment]
It is strange though that the default there is not Serializable.

File: spark/src/test/scala/org/apache/spark/sql/delta/WriteSuiteBase.scala [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+abstract class WriteSuiteBase
+  extends QueryTest
+  with SharedSparkSession
+  with BeforeAndAfterEach
+  with SQLTestUtils {
+
+  var tempDir: File = _
+
+  var deltaLog: DeltaLog = _
```
[the remark this answers is not in the document]
Comment (watfordkcf) [reply spark/src/test/scala/org/apache/spark/sql/delta/WriteSuiteBase.scala:37]: I copied from an existing test suite, I'll update.

File: spark/src/test/scala/org/apache/spark/sql/delta/WriteSuiteBase.scala [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  }
+
+  protected def readDeltaTableByPath(path: String): DataFrame = {
+    spark.read.format("delta").load(path)
+  }
+
+  override def beforeEach() {
+    super.beforeEach()
+    // Using a space in path to provide coverage for special characters.
+    tempDir = Utils.createTempDir(namePrefix = "spark test")
```
[the remark this answers is not in the document]
Comment (watfordkcf) [reply spark/src/test/scala/org/apache/spark/sql/delta/WriteSuiteBase.scala:52]: I copied from an existing test suite, I'll update.

File: spark/src/test/scala/org/apache/spark/sql/delta/WriteScalaSuite.scala [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+import org.apache.spark.sql.delta.sources.DeltaSQLConf
+import org.apache.spark.sql.delta.test.DeltaSQLCommandTest
+
+import org.apache.spark.sql.Row
+
+class WriteScalaSuite extends WriteSuiteBase with DeltaSQLCommandTest {
+
+  import testImplicits._
+
+  test("empty commits are elided on write by default") {
```
[the remark this answers is not in the document]
Comment (watfordkcf) [reply spark/src/test/scala/org/apache/spark/sql/delta/WriteScalaSuite.scala:28]: Sounds good, I just copied the existing pattern!

**Comment by vkorukanti** [comment]
> @prakharjain09 could you take a look at this?

ping @prakharjain09

**Comment by watfordkcf** [comment]
I've gone ahead and addressed the PR comments.

**Comment by watfordkcf** [comment]
Not sure what to do with the negative scalastyle result, the output isn't helpful at all:
https://github.com/delta-io/delta/actions/runs/5809827441/job/15749404064?pr=1934

**Comment by prakharjain09** [comment]
@vkorukanti Could you take a look at the failing `Delta Connectors Tests`? Any idea if the failing [Flink test](https://github.com/delta-io/delta/actions/runs/5858842742/job/15883596864?pr=1934) is flaky?

    [error] Test io.delta.flink.sink.DeltaSinkStreamingExecutionITCase.testFileSinkWithGlobalCommitterFailover failed: org.opentest4j.AssertionFailedError: expected: <45> but was: <49>, took 18.446s
    [error]     at io.delta.flink.sink.DeltaSinkStreamingExecutionITCase.runDeltaSinkTest(DeltaSinkStreamingExecutionITCase.java:519)
    [error]     at io.delta.flink.sink.DeltaSinkStreamingExecutionITCase.testFileSinkWithGlobalCommitterFailover(DeltaSinkStreamingExecutionITCase.java:241)
    [error]     ...

**Comment by vkorukanti** [comment]
> @vkorukanti Could you take a look at the failing `Delta Connectors Tests`? Any idea if the failing [Flink test](https://github.com/delta-io/delta/actions/runs/5858842742/job/15883596864?pr=1934) is flaky?
> 
> ```
> [error] Test io.delta.flink.sink.DeltaSinkStreamingExecutionITCase.testFileSinkWithGlobalCommitterFailover failed: org.opentest4j.AssertionFailedError: expected: <45> but was: <49>, took 18.446s
> [error]     at io.delta.flink.sink.DeltaSinkStreamingExecutionITCase.runDeltaSinkTest(DeltaSinkStreamingExecutionITCase.java:519)
> [error]     at io.delta.flink.sink.DeltaSinkStreamingExecutionITCase.testFileSinkWithGlobalCommitterFailover(DeltaSinkStreamingExecutionITCase.java:241)
> [error]     ...
> ```

This test is known to be flakey. cc. @scottsand-db

**Comment by watfordkcf** [comment]
My last day at KCF is 2023-09-13, I've subscribed to this PR from my personal GitHub and will attempt to follow-up there as needed.

**Comment by tdas** [comment]
oh no! we missed merging this. let me kick off the tests again.

**Pull request closed without merging**